修复manifestEqual方法缺陷

veraicon opened the pull request:
<!--  Thanks for submitting a pull request! Here are some tips for you:
1. Please make sure you have read and understood the contributing guidelines: https://github.com/AliyunContainerService/image-syncer/blob/main/CONTRIBUTING.md
2. Please make sure the PR has a corresponding issue.
-->

Pull Request Description
---

### Describe what this PR does / why we need it

修复manifestEqual方法缺陷。
该方法的意图应当是对比来源和目标镜像的manifest，如果相同就skip该任务。但是该方法之前在本地和远程manifest内容实际上完全一致，但是key顺序不同的情况下，会返回false，没有成功比较。
而由于destManifest是本地构造的，因此key顺序大概率与远程不同。
这导致重复向远程仓库提交push；而在远程仓库开启immutable的情况下，会导致任务预期外的失败。


### Does this pull request fix one issue?
<!--If that, add "Fixes #xxxx" below in the next line. For example, Fixes #15. Otherwise, add "NONE" -->
是的，已测试

### Describe how you did it

将简单的json.compact成为无序byte数组的比较，转变成递归的key value逐个值比较，包括slice。

### Describe how to verify it
将目标镜像仓库设置immutable，也就是已存在的tag不允许重复提交；此时将来源和目标设置为相同地址进行提交，在原始版本会报错。
unknown: The requested tag already exists and cannot be overwritten
当前提交版本会提示 skip synchronization because destination image exists 并返回成功

### Special notes for reviews

Commit:
修复manifestEqual方法缺陷。该方法之前在manifest内容实际上完全一致，但是key顺序不同的情况下，会返回false，没有成功比较。而由于destManifest是本地构造的，因此key顺序大概率与远程不同。这导致重复向远程仓库提交push；而在远程仓库开启immutable的情况下，会导致任务预期外的失败。

veraicon commented:
举例来说，根据日志，之前的版本会对比如下两个str，并返回false，实际上他们的内容值是一致的（这里对digest进行了mask）：
{
   "mediaType": "application/vnd.docker.distribution.manifest.list.v2+json",
   "schemaVersion": 2,
   "manifests": [
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "digest": "sha256:******",
         "size": 2831,
         "platform": {
            "architecture": "arm64",
            "os": "linux"
         }
      },
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "digest": "sha256:******",
         "size": 2831,
         "platform": {
            "architecture": "amd64",
            "os": "linux"
         }
      }
   ]
}



{"schemaVersion":2,"mediaType":"application/vnd.docker.distribution.manifest.list.v2+json","manifests":[{"mediaType":"application/vnd.docker.distribution.manifest.v2+json","size":2831,"digest":"sha256:******","platform":{"architecture":"arm64","os":"linux"}},{"mediaType":"application/vnd.docker.distribution.manifest.v2+json","size":2831,"digest":"sha256:******","platform":{"architecture":"amd64","os":"linux"}}]}

denverdino reviewed (commented):
```
        json.Unmarshal(m1, &a)
	json.Unmarshal(m2, &b)
```
Pls add error check for Unmarshal

And how about to use  ```eq := reflect.DeepEqual(m1, m2)```

More options for pkg maps and cmp in Golang 1.21

https://blog.stackademic.com/introducing-new-packages-in-go-maps-and-cmp-2639e64b6a85

Commit:
Add reflection error handling and replace custom implementation with deepEqual.

veraicon commented:
> ```
>         json.Unmarshal(m1, &a)
> 	json.Unmarshal(m2, &b)
> ```
> 
> Pls add error check for Unmarshal
> 
> And how about to use `eq := reflect.DeepEqual(m1, m2)`
> 
> More options for pkg maps and cmp in Golang 1.21
> 
> https://blog.stackademic.com/introducing-new-packages-in-go-maps-and-cmp-2639e64b6a85

Added unmarshal error handling and replaced custom implementation with  deepEqual.
Pls review ,thanks.
